hydra-task/src/main/java/com/addthis/hydra/task/pipeline/PipelineTask.java

```diff
@@ -0,0 +1,137 @@
+ * <p>A pipeline job will run the first job phase to completion, then the second
+ * phase, then the third phase, etc. When a pipeline job stops and is restarted
+ * it begins processing from the first phase. It is recommended that phases use
+ * the traditional mechanisms for processing data at most once, ie. the marks
+ * directories.</p>
+ * <p>Use the notation {@code ${hydra.task.jobid}} to retrieve the identifier
+ * of the currently running job. This should be used by downstream phases to read
+ * the files produced by an upstream phase.</p>
+ * <p>By default error checking is enabled to verify that each job phase
+ * does not write to an output directory of another job phase. If you want
+ * to disable this error checking then set {@code validateDirs} to false.</p>
+ * <p>All logging information is printed out using human (counting from 1)
+ * numbering of the phases.</p>
+ * <p>Example:</p>
+ * <pre>{
+ *    type : "pipeline",
+ *    phases: [
+ *    ]
+ * }</pre>
+ *
+ * @user-reference
+ * @hydra-name pipeline
+ */
+public class PipelineTask implements TaskRunnable {
+
+    private static final Logger log = LoggerFactory.getLogger(PipelineTask.class);
+
+    @Nonnull private final StreamMapper[] phases;
+
+    private final ImmutableList<CompletableFuture<Void>> phaseComplete;
+
+    private final ImmutableList<CompletableFuture<Void>> phaseNext;
+
+    private volatile StreamMapper currentPhase = null;
+
+    @JsonCreator
+    public PipelineTask(@JsonProperty("phases") @Nonnull StreamMapper[] phases) {
+        this.phases = phases;
+        int futures = Math.max(phases.length - 1, 0);
+        ImmutableList.Builder<CompletableFuture<Void>> complete = new ImmutableList.Builder<>();
+        ImmutableList.Builder<CompletableFuture<Void>> next = new ImmutableList.Builder<>();
+        for (int i = 0; i < futures; i++) {
+            final int current = i;
+            CompletableFuture<Void> onTaskComplete = phases[i].getOnTaskComplete();
+            CompletableFuture<Void> onTaskNext = onTaskComplete.thenRun(() -> beginPhase(current + 1));
+            complete.add(onTaskComplete);
+            next.add(onTaskNext);
+        }
+        this.phaseComplete = complete.build();
+        this.phaseNext = next.build();
+    }
+
+    @Override public void start() {
+        beginPhase(0);
+    }
+
+    @Override public void close() throws Exception {
+        log.info("Pipeline task is starting shutdown");
+        int size = phaseComplete.size();
+        for (int i = (size - 1); i >= 0; i--) {
+            phaseComplete.get(i).cancel(false);
+        }
+        for (int i = (size - 1); i >= 0; i--) {
+            try {
+                if (!phaseComplete.get(i).isCancelled()) {
+                    phaseNext.get(i).join();
+                }
+            } catch (CompletionException ex) {
+                log.error("Phase {} phaseNext future encountered an exception while starting phase {}: ",
```

Comment:
logging and then rethrowing errors is usually not a great sign, but the indexing information does seem helpful. Maybe this is a use case for wrapping exceptions -- eg. `new SomeExceptionType("Phase .... while starting phase...").